File: backend/routes/activities.py
from fastapi import APIRouter, HTTPException, Depends, status
from typing import List
from ..models.activity import Activity, ActivityCreate, ActivityUpdate
from motor.motor_asyncio import AsyncIOMotorDatabase
from ..database import get_database
import logging
from datetime import datetime

router = APIRouter(prefix="/activities", tags=["activities"])
logger = logging.getLogger(__name__)

@router.get("/", response_model=dict)
async def get_all_activities(db: AsyncIOMotorDatabase = Depends(get_database)):
    """Get all active activities ordered by week"""
    try:
        cursor = db.activities.find(
            {"is_active": True},
            sort=[("week", 1)]
        )
        activities = await cursor.to_list(length=100)
        
        # Convert ObjectId to string
        for activity in activities:
            activity["_id"] = str(activity["_id"])
            
        logger.info(f"Retrieved {len(activities)} activities")
        return {
            "success": True,
            "data": activities
        }
    except Exception as e:
        logger.error(f"Error fetching activities: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Error retrieving activities"
        )

@router.post("/", response_model=dict, status_code=status.HTTP_201_CREATED)
async def create_activity(
    activity_data: ActivityCreate,
    db: AsyncIOMotorDatabase = Depends(get_database)
):
    """Create a new activity"""
    try:
        # Check if week already exists
        existing = await db.activities.find_one({"week": activity_data.week, "is_active": True})
        if existing:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Activity for week {activity_data.week} already exists"
            )
        
        # Create new activity
        activity_dict = activity_data.model_dump()
        activity_dict["created_at"] = datetime.utcnow()
        activity_dict["updated_at"] = datetime.utcnow()
        
        result = await db.activities.insert_one(activity_dict)
        
        # Retrieve the created activity
        created_activity = await db.activities.find_one({"_id": result.inserted_id})
        created_activity["_id"] = str(created_activity["_id"])
        
        logger.info(f"Created activity for week {activity_data.week}")
        return {
            "success": True,
            "message": "Activity created successfully",
            "data": created_activity
        }
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error creating activity: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Error creating activity"
        )

@router.put("/{activity_id}", response_model=dict)
async def update_activity(
    activity_id: str,
    activity_data: ActivityUpdate,
    db: AsyncIOMotorDatabase = Depends(get_database)
):
    """Update an existing activity"""
    try:
        from bson import ObjectId
        
        # Check if activity exists
        existing = await db.activities.find_one({"_id": ObjectId(activity_id)})
        if not existing:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Activity not found"
            )
        
        # Prepare update data
        update_data = {k: v for k, v in activity_data.model_dump(exclude_unset=True).items()}
        
        if update_data:
            update_data["updated_at"] = datetime.utcnow()
            
            await db.activities.update_one(
                {"_id": ObjectId(activity_id)},
                {"$set": update_data}
            )
        
        # Return updated activity
        updated_activity = await db.activities.find_one({"_id": ObjectId(activity_id)})
        updated_activity["_id"] = str(updated_activity["_id"])
        
        logger.info(f"Updated activity {activity_id}")
        return {
            "success": True,
            "message": "Activity updated successfully",
            "data": updated_activity
        }
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error updating activity: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Error updating activity"
        )

@router.delete("/{activity_id}", response_model=dict)
async def delete_activity(
    activity_id: str,
    db: AsyncIOMotorDatabase = Depends(get_database)
):
    """Soft delete an activity (set is_active to False)"""
    try:
        from bson import ObjectId
        
        result = await db.activities.update_one(
            {"_id": ObjectId(activity_id)},
            {"$set": {"is_active": False, "updated_at": datetime.utcnow()}}
        )
        
        if result.matched_count == 0:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Activity not found"
            )
        
        logger.info(f"Deleted activity {activity_id}")
        return {
            "success": True,
            "message": "Activity deleted successfully"
        }
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error deleting activity: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Error deleting activity"
        )

File: backend-node/src/routes/activities.ts
import express from 'express';
import { body, param, validationResult } from 'express-validator';
import { Activity, IActivityCreate, IActivityUpdate } from '../models/Activity';
import { ApiResponse } from '../types/api';
import { asyncHandler } from '../middleware/errorHandler';

const router = express.Router();

// Validation middleware
const validateActivity = [
  body('week')
    .isInt({ min: 1, max: 5 })
    .withMessage('Week must be an integer between 1 and 5'),
  body('title')
    .isString()
    .trim()
    .isLength({ min: 1, max: 200 })
    .withMessage('Title must be between 1 and 200 characters'),
  body('description')
    .isString()
    .trim()
    .isLength({ min: 1, max: 1000 })
    .withMessage('Description must be between 1 and 1000 characters'),
  body('images')
    .optional()
    .isArray()
    .withMessage('Images must be an array of URLs'),
  body('images.*')
    .optional()
    .isURL()
    .withMessage('Each image must be a valid URL'),
];

const validateActivityUpdate = [
  body('week')
    .optional()
    .isInt({ min: 1, max: 5 })
    .withMessage('Week must be an integer between 1 and 5'),
  body('title')
    .optional()
    .isString()
    .trim()
    .isLength({ min: 1, max: 200 })
    .withMessage('Title must be between 1 and 200 characters'),
  body('description')
    .optional()
    .isString()
    .trim()
    .isLength({ min: 1, max: 1000 })
    .withMessage('Description must be between 1 and 1000 characters'),
  body('images')
    .optional()
    .isArray()
    .withMessage('Images must be an array of URLs'),
  body('images.*')
    .optional()
    .isURL()
    .withMessage('Each image must be a valid URL'),
  body('is_active')
    .optional()
    .isBoolean()
    .withMessage('is_active must be a boolean'),
];

// Helper function to handle validation errors
const handleValidationErrors = (req: express.Request, res: express.Response<ApiResponse>): boolean => {
  const errors = validationResult(req);
  if (!errors.isEmpty()) {
    res.status(400).json({
      success: false,
      message: 'Validation error',
      error: errors.array().map(err => err.msg).join(', '),
    });
    return true;
  }
  return false;
};

// GET /api/activities - Get all active activities
router.get('/', asyncHandler(async (req: express.Request, res: express.Response<ApiResponse>) => {
  try {
    const activities = await Activity.find({ is_active: true })
      .sort({ week: 1 })
      .limit(100)
      .lean();

    console.log(`📋 Retrieved ${activities.length} activities`);
    
    res.json({
      success: true,
      data: activities,
    });
  } catch (error) {
    console.error('❌ Error fetching activities:', error);
    res.status(500).json({
      success: false,
      message: 'Error retrieving activities',
    });
  }
}));

// POST /api/activities - Create new activity
router.post('/', validateActivity, asyncHandler(async (req: express.Request, res: express.Response<ApiResponse>): Promise<void> => {
  if (handleValidationErrors(req, res)) return;

  try {
    const activityData: IActivityCreate = req.body;

    // Check if week already exists
    const existingActivity = await Activity.findOne({ 
      week: activityData.week, 
      is_active: true 
    });

    if (existingActivity) {
      return res.status(400).json({
        success: false,
        message: `Activity for week ${activityData.week} already exists`,
      });
    }

    // Create new activity
    const activity = new Activity(activityData);
    await activity.save();

    console.log(`✅ Created activity for week ${activityData.week}`);

    res.status(201).json({
      success: true,
      message: 'Activity created successfully',
      data: activity.toObject(),
    });
  } catch (error) {
    console.error('❌ Error creating activity:', error);
    res.status(500).json({
      success: false,
      message: 'Error creating activity',
    });
  }
}));

// PUT /api/activities/:id - Update activity
router.put('/:id', 
  param('id').isString().withMessage('Activity ID must be a string'),
  validateActivityUpdate, 
  asyncHandler(async (req: express.Request, res: express.Response<ApiResponse>) => {
    if (handleValidationErrors(req, res)) return;

    try {
      const { id } = req.params;
      const updateData: IActivityUpdate = req.body;

      // Find and update activity
      const activity = await Activity.findOneAndUpdate(
        { id },
        { ...updateData, updated_at: new Date() },
        { new: true, runValidators: true }
      );

      if (!activity) {
        return res.status(404).json({
          success: false,
          message: 'Activity not found',
        });
      }

      console.log(`✅ Updated activity ${id}`);

      res.json({
        success: true,
        message: 'Activity updated successfully',
        data: activity.toObject(),
      });
    } catch (error) {
      console.error('❌ Error updating activity:', error);
      res.status(500).json({
        success: false,
        message: 'Error updating activity',
      });
    }
  })
);

// DELETE /api/activities/:id - Soft delete activity
router.delete('/:id',
  param('id').isString().withMessage('Activity ID must be a string'),
  asyncHandler(async (req: express.Request, res: express.Response<ApiResponse>) => {
    if (handleValidationErrors(req, res)) return;

    try {
      const { id } = req.params;

      const activity = await Activity.findOneAndUpdate(
        { id },
        { is_active: false, updated_at: new Date() },
        { new: true }
      );

      if (!activity) {
        return res.status(404).json({
          success: false,
          message: 'Activity not found',
        });
      }

      console.log(`🗑️ Deleted activity ${id}`);

      res.json({
        success: true,
        message: 'Activity deleted successfully',
      });
    } catch (error) {
      console.error('❌ Error deleting activity:', error);
      res.status(500).json({
        success: false,
        message: 'Error deleting activity',
      });
    }
  })
);

export default router;